File: day02/day02.js
var fs = require('fs');
const { listenerCount } = require('process');
let dataset = fs.readFileSync('day02_input.txt').toString().split("\n");

function range(s) {
    return s.split("-").map(x => parseInt(x))
}

dataset = dataset.map(x => x.split(' '))
dataset = dataset.map(x => [range(x[0]), x[1][0], x[2]])

// first item is number range
// second item is letter
// third item is password

// qn 1
function one(arr) {
    function char_count(s, char) {
        let count = 0
        for (let i = 0; i < s.length; i = i + 1) {
            if (char === s[i]) {
                count = count + 1
            }
            else { }
        }
        return count
    }

    let count = 0
    for (let i = 0; i < arr.length; i = i + 1) {
        let elem = arr[i]
        var left = elem[0][0]
        var right = elem[0][1]
        var letter = elem[1]
        var pwd = elem[2]
        var num = char_count(pwd, letter)

        if (num < left || num > right) { }
        else {
            count = count + 1
        }
    }
    return count
}

// qn 2
function two(arr) {
    let count = 0
    for (let i = 0; i < arr.length; i = i + 1) {
        let elem = arr[i]
        var left = elem[0][0] - 1
        var right = elem[0][1] - 1
        var letter = elem[1]
        var pwd = elem[2]
        
        if (pwd[left] === letter && pwd[right] === letter) {

        }
        else if (pwd[left] !== letter && pwd[right] !== letter) {

        }
        else {
            count = count + 1
        }
    }
    return count
}

console.log(one(dataset))
console.log(two(dataset))

File: day02/day02.py
dataset = open("day02_input.txt", 'r').read().split("\n")

# part 1
def parse_one(s):
    arr = s.split(" ")
    number = list(map(lambda x: int(x), arr[0].split("-")))
    letter = arr[1]
    pwd = arr[2]
    count = 0
    for i in pwd:
        if i == letter[0]:
            count += 1
    if count < number[0] or count > number[1]:
        return 0
    else:
        return 1


number = 0
for elem in dataset:
    number += parse_one(elem)
print(number)

# part 2
def parse_two(s):
    arr = s.split(" ")
    number = list(map(lambda x: int(x) - 1, arr[0].split("-")))
    letter = arr[1][0]
    pwd = arr[2]
    if ((pwd[number[0]] == letter and pwd[number[1]] != letter) 
        or (pwd[number[0]] != letter and pwd[number[1]] == letter)):
        return 1
    else:
        return 0

number = 0
for elem in dataset:
    number += parse_two(elem)
print(number)